src/core/qgsmaplayerregistry.cpp

```diff
@@ -203,14 +220,17 @@ void QgsMapLayerRegistry::onMapLayerDeleted( QObject* obj )
 
 QMap<QString, QgsMapLayer*> QgsMapLayerRegistry::mapLayers()
 {
-  return mMapLayers;
+  QReadLocker lock( &mRegistryLock );
+
+  QMap<QString, QgsMapLayer*> mapLayers = mMapLayers;
+  mapLayers.detach();
+  return mapLayers;
```

Comment:
As far as I understand that, the fact that you have an implicitly shared class does not guarantee thread safety. E.g. if you have one instance and two references to it, one in each thread, implicit sharing will not help - you still need locking. But if you have two instances (sharing the same data internally), one in each thread, then you are fine, because whenever needed, detach() will be called - and the instances will stop sharing the data.